File: packages/sdk/java/packages/admin/src/main/java/dev/edgebase/sdk/admin/AnalyticsClient.java
package dev.edgebase.sdk.admin;

import dev.edgebase.sdk.core.EdgeBaseError;
import dev.edgebase.sdk.core.generated.GeneratedClientWrappers;
import dev.edgebase.sdk.core.generated.GeneratedDbApi;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class AnalyticsClient {
    public static class AnalyticsEvent {
        public final String name;
        public final Map<String, ?> properties;
        public final Long timestamp;
        public final String userId;

        public AnalyticsEvent(String name) {
            this(name, Collections.emptyMap(), null, null);
        }

        public AnalyticsEvent(String name, Map<String, ?> properties, Long timestamp, String userId) {
            this.name = name;
            this.properties = properties;
            this.timestamp = timestamp;
            this.userId = userId;
        }
    }

    private final GeneratedClientWrappers.AnalyticsMethods methods;
    private final dev.edgebase.sdk.admin.generated.GeneratedAdminApi adminCore;

    public AnalyticsClient(GeneratedDbApi core, dev.edgebase.sdk.admin.generated.GeneratedAdminApi adminCore) {
        this.methods = new GeneratedClientWrappers.AnalyticsMethods(core);
        this.adminCore = adminCore;
    }

    @SuppressWarnings("unchecked")
    public Map<String, Object> overview() throws EdgeBaseError {
        Object result = adminCore.queryAnalytics(Map.of("metric", "overview"));
        return result instanceof Map<?, ?> map ? (Map<String, Object>) map : Collections.emptyMap();
    }

    @SuppressWarnings("unchecked")
    public Map<String, Object> overview(Map<String, String> options) throws EdgeBaseError {
        Object result = adminCore.queryAnalytics(buildQuery("overview", options));
        return result instanceof Map<?, ?> map ? (Map<String, Object>) map : Collections.emptyMap();
    }

    @SuppressWarnings("unchecked")
    public List<Map<String, Object>> timeSeries() throws EdgeBaseError {
        return timeSeries(Collections.emptyMap());
    }

    @SuppressWarnings("unchecked")
    public List<Map<String, Object>> timeSeries(Map<String, String> options) throws EdgeBaseError {
        Object result = adminCore.queryAnalytics(buildQuery("timeSeries", options));
        if (result instanceof Map<?, ?> map && map.get("timeSeries") instanceof List<?> list) {
            return (List<Map<String, Object>>) list;
        }
        return Collections.emptyList();
    }

    @SuppressWarnings("unchecked")
    public List<Map<String, Object>> breakdown() throws EdgeBaseError {
        return breakdown(Collections.emptyMap());
    }

    @SuppressWarnings("unchecked")
    public List<Map<String, Object>> breakdown(Map<String, String> options) throws EdgeBaseError {
        Object result = adminCore.queryAnalytics(buildQuery("breakdown", options));
        if (result instanceof Map<?, ?> map && map.get("breakdown") instanceof List<?> list) {
            return (List<Map<String, Object>>) list;
        }
        return Collections.emptyList();
    }

    @SuppressWarnings("unchecked")
    public List<Map<String, Object>> topEndpoints() throws EdgeBaseError {
        return topEndpoints(Collections.emptyMap());
    }

    @SuppressWarnings("unchecked")
    public List<Map<String, Object>> topEndpoints(Map<String, String> options) throws EdgeBaseError {
        Object result = adminCore.queryAnalytics(buildQuery("topEndpoints", options));
        if (result instanceof Map<?, ?> map && map.get("topItems") instanceof List<?> list) {
            return (List<Map<String, Object>>) list;
        }
        return Collections.emptyList();
    }

    public void track(String name) throws EdgeBaseError {
        track(name, Collections.emptyMap(), null);
    }

    public void track(String name, Map<String, ?> properties) throws EdgeBaseError {
        track(name, properties, null);
    }

    public void track(String name, Map<String, ?> properties, String userId) throws EdgeBaseError {
        trackBatch(Collections.singletonList(new AnalyticsEvent(name, properties, null, userId)));
    }

    public void trackBatch(List<AnalyticsEvent> events) throws EdgeBaseError {
        if (events == null || events.isEmpty()) {
            return;
        }

        List<Map<String, Object>> payloadEvents = new ArrayList<>();
        for (AnalyticsEvent event : events) {
            Map<String, Object> payload = new LinkedHashMap<>();
            payload.put("name", event.name);
            if (event.properties != null && !event.properties.isEmpty()) {
                payload.put("properties", event.properties);
            }
            if (event.userId != null && !event.userId.isEmpty()) {
                payload.put("userId", event.userId);
            }
            payload.put("timestamp", event.timestamp != null ? event.timestamp : System.currentTimeMillis());
            payloadEvents.add(payload);
        }

        Map<String, Object> body = new LinkedHashMap<>();
        body.put("events", payloadEvents);
        methods.track(body);
    }

    public Object queryEvents() throws EdgeBaseError {
        return queryEvents(Collections.emptyMap());
    }

    public Object queryEvents(Map<String, String> options) throws EdgeBaseError {
        return adminCore.queryCustomEvents(options);
    }

    public void destroy() {
        // Admin analytics calls are request-scoped; destroy is a compatibility no-op.
    }

    private Map<String, String> buildQuery(String metric, Map<String, String> options) {
        Map<String, String> query = new LinkedHashMap<>();
        query.put("metric", metric);
        if (options != null) {
            query.putAll(options);
        }
        return query;
    }
}

File: packages/sdk/kotlin/admin/src/main/kotlin/dev/edgebase/sdk/admin/AnalyticsClient.kt
package dev.edgebase.sdk.admin

import dev.edgebase.sdk.admin.generated.GeneratedAdminApi
import dev.edgebase.sdk.core.currentTimeMillis
import dev.edgebase.sdk.core.generated.GeneratedAnalyticsMethods
import dev.edgebase.sdk.core.generated.GeneratedDbApi

data class AnalyticsEvent(
    val name: String,
    val properties: Map<String, Any?>? = null,
    val timestamp: Long? = null,
    val userId: String? = null,
)

class AnalyticsClient(
    core: GeneratedDbApi,
    private val adminCore: GeneratedAdminApi,
) {
    private val methods = GeneratedAnalyticsMethods(core)

    @Suppress("UNCHECKED_CAST")
    suspend fun overview(options: Map<String, String> = emptyMap()): Map<String, Any?> =
        adminCore.queryAnalytics(buildQuery("overview", options)) as? Map<String, Any?> ?: emptyMap()

    @Suppress("UNCHECKED_CAST")
    suspend fun timeSeries(options: Map<String, String> = emptyMap()): List<Map<String, Any?>> {
        val result = adminCore.queryAnalytics(buildQuery("timeSeries", options)) as? Map<String, Any?> ?: return emptyList()
        return result["timeSeries"] as? List<Map<String, Any?>> ?: emptyList()
    }

    @Suppress("UNCHECKED_CAST")
    suspend fun breakdown(options: Map<String, String> = emptyMap()): List<Map<String, Any?>> {
        val result = adminCore.queryAnalytics(buildQuery("breakdown", options)) as? Map<String, Any?> ?: return emptyList()
        return result["breakdown"] as? List<Map<String, Any?>> ?: emptyList()
    }

    @Suppress("UNCHECKED_CAST")
    suspend fun topEndpoints(options: Map<String, String> = emptyMap()): List<Map<String, Any?>> {
        val result = adminCore.queryAnalytics(buildQuery("topEndpoints", options)) as? Map<String, Any?> ?: return emptyList()
        return result["topItems"] as? List<Map<String, Any?>> ?: emptyList()
    }

    suspend fun track(name: String, properties: Map<String, Any?> = emptyMap(), userId: String? = null) {
        trackBatch(listOf(AnalyticsEvent(name = name, properties = properties, userId = userId)))
    }

    suspend fun trackBatch(events: List<AnalyticsEvent>) {
        if (events.isEmpty()) return
        methods.track(
            mapOf(
                "events" to events.map { event ->
                    buildMap<String, Any?> {
                        put("name", event.name)
                        if (!event.properties.isNullOrEmpty()) put("properties", event.properties)
                        if (!event.userId.isNullOrBlank()) put("userId", event.userId)
                        put("timestamp", event.timestamp ?: currentTimeMillis())
                    }
                }
            )
        )
    }

    suspend fun queryEvents(options: Map<String, String> = emptyMap()): Any? =
        adminCore.queryCustomEvents(options.ifEmpty { null })

    private fun buildQuery(metric: String, options: Map<String, String>): Map<String, String> =
        buildMap {
            put("metric", metric)
            putAll(options)
        }
}
